Add topology visualization

tttttangTH opened the pull request:
Server and frontend implementation for topology visualization.

Commit:
Add rest HTTP server

Commit:
Add rest HTTP server

Commit:
Add 3rd party library cJSON into third-party directory

Commit:
Add 3rd party library http-parser into third-party directory

Commit:
Add 3rd party library d3js into third-party directory

Commit:
Add topology functionality

Commit:
Add topology functionality

Commit:
Add topology display in HTML file.

Commit:
Add topology display in HTML file.

Commit:
Add style for topology visualization

Commit:
Add style for topology visualization

Commit:
Add library d3js to subdirectory

Commit:
Add Rest server to otbr-agent mainloop

Commit:
Add Rest server to otbr-agent mainloop

Commit:
Add Rest server and 3rd party library cJSON and http-parser to CMAKE file

wgtdkp commented on src/agent/main.cpp:
```diff
@@ -103,6 +105,10 @@ static int Mainloop(otbr::AgentInstance &aInstance, const char *aInterfaceName)
 #else
     (void)aInterfaceName;
 #endif
+    ControllerOpenThread *     ncpOpenThread1 = reinterpret_cast<ControllerOpenThread *>(&aInstance.GetNcp());
```
is `ncpOpenThread1` necessary?

tttttangTH replied:
Does rest_web_server need its own if clause to control its initialization? It seem not reasonable if I put it within "#if OTBR_ENABLE_DBUS_SERVER"

wgtdkp replied:
okay, makes sense. But please have a good name.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+#include <netinet/in.h>
```
add copyright doc at the beginning.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+#include <netinet/in.h>
+#include <stdio.h>
```
sort the headers.
For header inclusion order, please reference to https://github.com/openthread/openthread/blob/master/STYLE_GUIDE.md.

And generally, you style needs to conform to this style guide.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+namespace agent {
+
+
+static int on_message_begin(http_parser *parser) {
```
Follow OpenThread naming style:
```suggestion
static int OnMessageBegin(http_parser *parser) {
```

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+static int on_status(http_parser *parser,const char *at, size_t len) {
+    Connection *Connection = (struct Connection*)parser->data;
+    Connection->status = new char[len+1]();
+    strncpy(Connection->status,at,len);
```
add space after comma `,`.

You should run `./script/make-pretty clang` to format your code.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+	return 0;
+}
+
+static http_parser_settings mSettings {
```
We use `m` monation for only member fields of class/struct. 
```suggestion
static http_parser_settings sSettings {
```

`s` stands for `static`.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    
+    handlerInit();
+    
+    mAddress->sin_family = AF_INET;
```
Does this support IPv6? For example, if a user request to this border router by its Ipv6 address, will this work?

tttttangTH replied:
Currently it can't. Could modify the " socket_in mAddress" from "socket_in"  to  "socket_in6", then IPv6 is supported. Will investigate on it later.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    mAddress->sin_addr.s_addr = INADDR_ANY;
+    mAddress->sin_port = htons(80);
+
+    mListenFd = socket(AF_INET, SOCK_STREAM, 0);
```
check return value.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    mListenFd = socket(AF_INET, SOCK_STREAM, 0);
+    otbrLog(OTBR_LOG_ERR, "socket success" );
+    
+    bind(mListenFd, (struct sockaddr *) mAddress, sizeof(sockaddr)) ;
```
dito.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    bind(mListenFd, (struct sockaddr *) mAddress, sizeof(sockaddr)) ;
+    otbrLog(OTBR_LOG_ERR, "bind success" );
+    
+    listen(mListenFd, 5);
```
dito.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    listen(mListenFd, 5);
+    otbrLog(OTBR_LOG_ERR, "listen success" );
+    
+    setNonBlocking(mListenFd);
```
```suggestion
    SetNonBlocking(mListenFd);
```

Per OpenThread style guide, a function name should start with uppercase.

wgtdkp replied:
Please re-check for anywhere else.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+ void RestWebServer::UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd,timeval &aTimeout){
+
+        
+        struct timeval timeout        = {60*60*24,0};
```
I think there is already a constant value for this purpose.

tttttangTH replied:
Is it the "kPollTimeout" in main.cpp of otbr-agent?
I have already initialized it in another approach.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+                timeout.tv_sec = 0; 
+                timeout.tv_usec = 0; 
+            }
+            if(it->second->diagisRequest){
```
`diagisRequest` is a typo?

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+        
+        for ( auto it = mConnectionSet.begin(); it != mConnectionSet.end(); ++it ){
+        
+        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - it->second->start).count();
```
What about give a more descriptive name to `it->second`.  Because it is not readable as I don't know what `it->second` is without check the definition of the ConnectionSet by myself.

So I suggest:
```C++
Foo &foo = it->second;
```
and reference to the name `foo` anywhere else.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    write(connection->fd, h1, strlen(h1));
+    write(connection->fd, h3, strlen(h3));
+    write(connection->fd, "\r\n", 2);
+    write(connection->fd,p, strlen(p));
```
We should define short functions for this.

for example,
```C++
int Connection::WriteResponseHeader(int aCode, const char *aMessage);
int Connection::WriteContentType(const char *aContentType);
```

And we should buffer the response data and send once at the last, instead of calling `write(...)` for each data segment.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    
+    free(p);
+
+    // cJSON_Delete(data);
```
If you are sure that this should be removed, then remove it.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    return (((v & 0x00ffU) << 8) & 0xff00) | (((v & 0xff00U) >> 8) & 0x00ff);
+}
+
+char* RestWebServer::formatBytes(const uint8_t *aBytes, uint8_t aLength){
```
is there already function in otbr/common that does converting integer to hex string?

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    
+ }
+
+cJSON* RestWebServer::getJsonState(Connection* connection,otInstance * mInstance){
```
```suggestion
cJSON* RestWebServer::getJsonState(Connection* connection,otInstance * aInstance){
```

And everywhere else.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    const char *networkName = otThreadGetNetworkName(mInstance);
+    char *p = new char[20]();
+    sprintf(p,"%s",static_cast<const char *>(networkName));
+    cJSON* ret = cJSON_CreateString(p);
```
Will `p` be auto-freed by `cJSON`?

tttttangTH replied:
No, need to be delete manually.

tttttangTH replied:
Will do that.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    uint16_t rloc16 = otThreadGetRloc16(mInstance);
+    char *p = new char[7]();
+    sprintf(p,"0x%04x",rloc16);
+    cJSON * ret = cJSON_CreateString(p);
```
I think we can create an Integer JSON value for this, right?

tttttangTH replied:
Yes, currently,  for simplicity, we packet everything as string and process it on the frontend side. It is easier to understand the data if we create object as what it is.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    char *p = new char[17]();
+    char *q = p;
+    
+    for (int i = 0; i <  OT_EXT_PAN_ID_SIZE; i++)
```
Write a `Bytes2HexString` function if there is no in otbr/common.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    
+    connection->hasCallback = 1;
+    connection->diagisRequest = 1;
+    connection->diagInfo = new DiagInfo(std::chrono::steady_clock::now());
```
is `diagInfo` necessary to be a pointer?

wgtdkp replied:
Generally
1. we should avoid dynamic allocation if possible.
2. If dynamic allocation is necessary, prefer `std::unique_ptr`/`std::shared_ptr` against raw pointer.

wgtdkp commented on src/agent/rest_web_server.hpp:
```diff
@@ -0,0 +1,162 @@
+    DiagInfo(std::chrono::steady_clock::time_point astart):start(astart){}
+}DiagInfo;
+
+typedef struct Connection {
```
Shouldn't this be a `class` and hide the implementation details?

wgtdkp commented on src/agent/rest_web_server.hpp:
```diff
@@ -0,0 +1,162 @@
+    
+    cJSON* getJSON(int fd);
+
+    void FreeConnection(int fd);
```
I would expect this function accepts a parameter of `Connection*`.

wgtdkp commented on src/agent/rest_web_server.hpp:
```diff
@@ -0,0 +1,162 @@
+    std::unordered_map<int,Connection*> mConnectionSet;
+    
+    
+    // static DiagInfo* mDiagInfo;
```
delete it?

wgtdkp commented on src/agent/rest_web_server.hpp:
```diff
@@ -0,0 +1,162 @@
+    // static DiagInfo* mDiagInfo;
+    
+
+    //Handler
```
```suggestion
    // Handlers
```

wgtdkp commented on src/agent/rest_web_server.hpp:
```diff
@@ -0,0 +1,162 @@
+    static void diagnosticResponseHandler(otMessage *aMessage, const otMessageInfo *aMessageInfo, void *aContext);
+    void diagnosticResponseHandler(otMessage *aMessage, const otMessageInfo &aMessageInfo);
+    
+    static cJSON*  getJsonNodeInfo(Connection* connection, otInstance * mInstance);
```
Should those json functions go to a separate `Json` module?

tttttangTH replied:
What I want to implement is that these function are `Handler` module that can be mapped to each RESTful API,  although they return pointers to  a cJSON object.  Those `GetJson` functions call a Openthread API and reformat the return value as JSON then return, so it seems better that if we stripe all codes about formatting JSON then put them into a `JSON` module then  use them after getting the result of `Handler` module.

wgtdkp commented on src/agent/rest_web_server.cpp:
```diff
@@ -0,0 +1,1061 @@
+    return ret;
+
+}
+cJSON*  RestWebServer::getJsonRloc16(Connection* connection, otInstance * mInstance){
```
Do we need the argument `connection`?

tttttangTH replied:
Yes, need to set some parameters about connection

wgtdkp commented on src/web/web-service/frontend/index.html:
```diff
@@ -28,303 +28,455 @@
-
-  <!-- SEO: If your mobile URL is different from the desktop URL, add a canonical link to the desktop page https://developers.google.com/webmasters/smartphone-sites/feature-phones -->
-  <!--
+    <meta charset="utf-8">
```
why bother changing the indention size? It introduce unnecessary diff. Please keep the indent size as it is.

wgtdkp commented on src/web/web-service/frontend/res/css/styles.css:
```diff
@@ -14,231 +14,416 @@
 @keyframes spin {
-  0% { transform: rotate(0deg); }
-  100% { transform: rotate(360deg); }
+    0% {
```
Don't change previous format, even it is bad/wrong. We may change it in separate PR, but it is out of the scope of our project and this PR.

wgtdkp reviewed (commented):
1. remove `.DS_Store` files.
2. move third_parties into a `repo` sundirectory. for example move `cJson` from `third_party/cJson` to `third_party/cjson/repo`.

Commit:
Apply clang-format for rest_web_server

Commit:
Apply clang-format for rest_web_server

Commit:
combine intialization of rest_web_server with dbus

Commit:
combine intialization of rest_web_server with dbus

Commit:
modify link method of cJSON and http-parser

Commit:
reformat indent of index file

Commit:
reformat indent of index file

Commit:
modify click and mouseover event of topology

Commit:
modify click and mouseover event of topology

Commit:
remove DS_store file in cJSON

Commit:
remove .DS_store file in http-parser

Commit:
modify click and mouseover events in topology visualization

Commit:
modify click and mouseover events in topology visualization

Commit:
sort the headers

Commit:
sort the headers

Commit:
modify CMake configuration to include third party library

Commit:
modify init function name of rest server

Commit:
modify init function name of rest server

Commit:
Rename function name and parameter name, separate rest server into different module

Commit:
Rename function name and parameter name, separate rest server into different module

Commit:
remove third_party cJSON on remote repositories

Commit:
remove third party http-parser on remote repositoriy

Commit:
remove third party d3js on remote repository

Commit:
delete useless indent and modification.

Commit:
delete useless indent and modification.

Commit:
reformat index file

Commit:
reformat index file

Commit:
separate rest server into different module into src/rest directory

Commit:
separate rest server into different module into src/rest directory

Commit:
remove indent

Commit:
remove indent

Commit:
apply make-pretty check

Commit:
apply make-pretty check

Commit:
remove rest server in src/agent

Commit:
remove rest server in src/agent

Commit:
change rest mudule name

Commit:
change rest mudule name

Commit:
put rest out of the control of dbus

Commit:
put rest out of the control of dbus

Commit:
change rest cmake configuration

Commit:
change rest cmake configuration

Commit:
reorganize connection mudule in rest

Commit:
reorganize connection mudule in rest

Commit:
reorganize request mudule in rest

Commit:
reorganize request mudule in rest

Commit:
reorganize  resoponse mudule in rest

Commit:
reorganize  resoponse mudule in rest

Commit:
reorganize server  mudule in rest

Commit:
reorganize server  mudule in rest

Commit:
rename hanlder module as resource

Commit:
rename hanlder module as resource

Commit:
reorganize json module in rest

Commit:
reorganize json module in rest

Commit:
use '' instead of "" in app.js

Commit:
use '' instead of "" in app.js

Commit:
add third party library cjson and http-parser into third-party cmake file

Commit:
add third party library cjson and http-parser into third-party cmake file

Commit:
remove useless code

Commit:
remove useless code

Commit:
apply make pretty

Commit:
apply make pretty

tttttangTH commented on src/rest/connection.hpp:
```diff
@@ -0,0 +1,143 @@
+
+    bool IsReadTimeout(int aDuration);
+
+    bool IsCallbackTimeout(int aDuration)
```
miss ";" here

Commit:
remove empty line in main.cpp in otbr-agent

Commit:
remove empty line in main.cpp in otbr-agent

Commit:
Resolve comments and optimize structure

Commit:
Resolve comments and optimize structure

Commit:
add error check for rest server

Commit:
add error check for rest server

Commit:
remove hardcoding of hostname

Commit:
remove hardcoding of hostname

Commit:
resolve comments and modify logic for read and write

Commit:
modify parameters that rest server need in mainloop

Commit:
Merge branch 'manager-draft' of github.com:tttttangTH/ot-br-posix into manager-draft

Commit:
remove multiple returns in one function

Commit:
modify rest server port number set in frontend

Commit:
delete useless lines

Commit:
add first_party library  otbr-utils

Commit:
reslove comments and reorganize code

Commit:
add cmake configuration for third party library cJson

Commit:
add cmake configuration for third party library http-parser

Commit:
reorganize modules and fix bugs in accept

Commit:
add comments and make modifications according to the change of restful api

Commit:
add test for rest server

Commit:
Merge branch 'master' into manager-draft

Commit:
apply make-pretty

Commit:
add third party d3js cmake file

Commit:
fix bugs in frontend display

Commit:
add copyright

Commit:
remove useless include header file

Commit:
add third party d3js to third_party directory

Commit:
add third party cJSON to third_party library

Commit:
add third party library http-parser to third party directory

Commit:
apply make-pretty

Commit:
fix warning

Commit:
apply make pretty

Commit:
apply make-pretty

Commit:
add copyright

Commit:
remove unused variable

Commit:
put usage of http-parser in source file of module parser

Commit:
modify cmake configure file according to the changes in parser module

Commit:
set custom compile option off in cJSON

Commit:
modify http-parser cmake file according to modifications on parser module

Commit:
fix bugs in scan build

Commit:
fix bugs about undefined reference of rest server function

Commit:
fix warning in scan build

Commit:
put rest server into a wrap

Commit:
modify rest cmake file

Commit:
add rest source files into android.mk

Commit:
add header file for errno

Commit:
add cerrno to fix undefinition of errno

Commit:
try to fix errno error in android

Commit:
try to fix bug in android check about third_party library

Commit:
try to fix bug in android check

Commit:
add android configure file

Commit:
run all test of rest-test

Commit:
modify android configure file for third party cJSON and http-parser

Commit:
put third party target in top layer android configuration

Commit:
change file location

Commit:
modify third_party android configure file

Commit:
try to fix android build fail

Commit:
put third_party library in subdiretory

Commit:
put all module in top layer android configure file

Commit:
add local path prefix

Commit:
modify rest server as singleton pattern

Commit:
remove third_party for review

Commit:
add build option for otbr rest

Commit:
add otbr-rest option to github action

Commit:
resolve comments and add more http response code for response module

Commit:
add otbr-rest option

Commit:
delete uesless line

Commit:
modify rest server name

Commit:
fix typo

Commit:
Merge branch 'manager' into manager-draft

Commit:
remove useless file

Commit:
add rest in github action

Commit:
apply make-pretty

Commit:
fix unbound variable

Commit:
apply make pretty

Commit:
apply make-pretty check

Commit:
apply make-pretty

Commit:
modify rest test

Commit:
try to fix test bug

Commit:
fix bugs

Commit:
try to fix test bugs

Commit:
try to fix test issue

Commit:
fix test error

Commit:
fix test issue

Commit:
rename error check

Commit:
Merge branch 'manager' into manager-draft

Commit:
fix typo

Commit:
resolve comments

Commit:
apply make-pretty

Commit:
Merge branch 'manager' into manager-draft

Commit:
add rest option to github action

Commit:
Merge branch 'manager' into manager-draft

Commit:
resolve comments and detail modification

Commit:
add a new check for rest server

Commit:
Merge branch 'manager' into manager-draft

Commit:
change NULL to nullptr

Commit:
add otbr-web API

Commit:
Merge branch 'master' into manager

Commit:
add api for otbr-web

Commit:
modify frontend

Commit:
Merge branch 'manager-web' into manager

Commit:
add comments

Commit:
modify otbr-rest

Commit:
modify cJSON

Commit:
modify http-parser

Commit:
modify d3js

Commit:
detail modify

Commit:
apply make-pretty

Commit:
Merge branch 'manager' into manager-draft

Commit:
delete d3js library

Commit:
small modify

Commit:
modify frontend

Commit:
frontend display

Commit:
switch d3js version to v3.5.17

Commit:
add submodule cJSON

Commit:
add submodule http-parser

Commit:
modify frontend

Commit:
frontend modify

Commit:
fix build fail

Commit:
format file

Commit:
modify test for otbr-rest

Commit:
fix bug

Commit:
modify gitmodules

Commit:
delete three library

Commit:
add cJSON

Commit:
add http-parser

Commit:
add d3js

Commit:
remove useless file

Commit:
resolve comments

Commit:
switch to python3

Commit:
apply make-pretty

Commit:
address LGTM alert

Commit:
modify test for rest

Commit:
address lgtm alert

Commit:
add node check

Commit:
fix test bug

Commit:
modify CMakeList.txt file

Commit:
resolve comments

Commit:
fix package check fail

Commit:
fix build fail

bukepo committed:
[cmake] support build with 3.10 (#547)

Commit:
[web] unify MasterKey and NetworkKey (#546)

This commit changes all NetworkKey to MasterKey in Web module, so the
naming of this terminology is consistent with the Thread
Specification.

tttttangTH committed:
rename variable

tttttangTH committed:
Merge branch 'master' of github.com:openthread/ot-br-posix

tttttangTH committed:
Merge branch 'master' into manager-draft

tttttangTH committed:
remove dead code
